neuralmonkey/encoders/cnn_encoder.py

```diff
@@ -139,17 +255,111 @@ def output(self) -> tf.Tensor:
             dropout_keep_prob=self.dropout_keep_prob,
             train_mode=self.train_mode)
 
-    @tensor
-    def spatial_mask(self) -> tf.Tensor:
-        return tf.ones(tf.shape(self.spatial_states)[:3])
-
     def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
         # if it is from the pickled file, it is list, not numpy tensor,
         # so convert it as as a prevention
         images = np.array(dataset.get_series(self.data_id))
 
         f_dict = {}
         f_dict[self.image_input] = images / 225.0
+        # it is one everywhere where non-zero, i.e. zero columns are masked out
+        f_dict[self.image_mask] = np.sum(
+            np.sign(images), axis=3, keepdims=True)
 
         f_dict[self.train_mode] = train
         return f_dict
+
+
+def max_pooling(
+        prev_layer: tf.Tensor,
+        prev_mask: tf.Tensor,
+        specification: Tuple[str, int, int, str],
+        layer_num: int) -> Tuple[tf.Tensor, tf.Tensor]:
+    if len(specification) != 4:
```

Comment:
tohle vyřeší check_argument_types, ne?

Reply:
Mohl by, ale musel by se připsat explicitně do těhle metod.

Reply:
No já bych to sem připsal - je to type check ne value check. Ani by se teda neměl házet ValueError ale TypeError. Kontrolovat délku u tuplu je podlě mě nevhodný - tuple má na rozdíl od seznamu předem definovanou délku

a pokud se to nevolá z konfiguráku a věříme mypy, tak to tady neni potřeba vůbec, protože by to mělo chytit už mypy, že to nejde nakástovat.

Reply:
Mypy to nechytí, protože je tam `Union` a pak je tam `if ... elif ... elif` a castuje se to podle těch ifů, takže bych tam ty `check_argument_types` přidal.

Reply:
přidej a tim pádem tenhle if vyhoď. Trochu je mi ale líto tý informace o tom, ve kterým layeru to je. Jinak proč vlastně nepoužíváme `@typechecked`? Přijde mi to hezčí.

Reply:
Počkej a vleze sem ta chyba z konfiguráku nebo z kódu nebo z obojího?

Reply:
(btw. `@typechecked` by muselo bejt `@typechecked(always=True)`)

Reply:
Když je ten tuple napsanej blbě (ale ne dost blbě, aby se to poznalo dřív) v konfiguráku, tak doputuje až sem.

Reply:
Tak bych tam dal check_argument_types a případně ho obalil do try-catch bloku, kde se to reraisuje  s upravenou message ve kterým je to patře